internal/runners/ppm/shim.go

```diff
@@ -0,0 +1,107 @@
+package ppm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/ActiveState/cli/internal/analytics"
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/internal/runbits"
+	"github.com/ActiveState/cli/pkg/projectfile"
+)
+
+type Shim struct {
+	out         output.Outputer
+	intercepted string
+	replaced    string
+	localeID    string
+}
+
+func NewShim(out output.Outputer) *Shim {
+	return &Shim{
+		out: out,
+	}
+}
+
+func (s *Shim) RunPPM(args ...string) error {
+	s.intercepted = "ppm"
+	s.replaced = "packages"
+	s.localeID = "ppm_print_forward"
+	return s.shim(args...)
+}
+
+func (s *Shim) RunInstall(args ...string) error {
+	s.intercepted = "install"
+	s.replaced = "packages add"
+	s.localeID = "ppm_print_forward_failure"
+	return s.shim(args...)
+}
+
+func (s *Shim) RunUpgrade(args ...string) error {
+	s.intercepted = "upgrade"
+	s.replaced = "packages update"
+	s.localeID = "ppm_print_forward_failure"
+	return s.shim(args...)
+}
+
+func (s *Shim) RunRemove(args ...string) error {
+	s.intercepted = "remove"
+	s.replaced = "packages remove"
+	s.localeID = "ppm_print_forward_failure"
+	return s.shim(args...)
+}
+
+func (s *Shim) RunList(args ...string) error {
+	s.intercepted = "list"
+	s.replaced = "packages"
+	s.localeID = "ppm_print_forward"
+	return s.shim(args...)
+}
+
+func (s *Shim) shim(args ...string) error {
+	err := s.executeShim(args...)
+	if err != nil {
+		analytics.EventWithLabel(analytics.CatPPMShimCmd, "error", errs.Join(err, " :: ").Error())
+	} else {
+		analytics.EventWithLabel(analytics.CatPPMShimCmd, "success", fmt.Sprintf("intercepted=%s, replaced=%s", s.intercepted, s.replaced))
+	}
+	return err
+}
```

Comment:
Please get rid of this function and handle your execution and analytics directly from the invoker (one level higher). Each invocation should have its own analytics anyway, since right now you are loosing what the particular subcommand was.

Suggest:

category: CatPPMShimCmd
event: subcommand(s)
label: success | "error: %v" % errs.join()..

Reply:
Each invocation does have its own analytics, at least in terms of what's reported. The subcommand for ppm is captured in `s.intercepted`, which is now an argument.